datafusion/core/src/datasource/file_format/write/demux.rs

```diff
@@ -320,9 +324,11 @@ async fn hive_style_partitions_demuxer(
 fn compute_partition_keys_by_row<'a>(
     rb: &'a RecordBatch,
     partition_by: &'a [(String, DataType)],
-) -> Result<Vec<Vec<&'a str>>> {
+) -> Result<Vec<Vec<String>>> {
```

Comment:
🤔  I wonder if computing new strings for each row will be unnecessarily slow 🤔  The current code only allocates a string for each distinct partition value (in the final take map) but this code now creates a new string for each row in the output record batch, just to match them up

Reply:
I vaguely recall I had to make this change because of an issue with borrowing temp values with the dates but I'd have to switch back to see exactly where the cause was. You are obviously correct though in that there is a fairly obvious overhead to using String. I'll take another look and see if there is something I can come up with.

Reply:
Maybe  you could at least use [`Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html) to avoid causing a regression for `StringArrays` 

I think it would be fine if existing functionality was unaffected but new features (aka partitioning on newly supported types) was not as fast as it could be

We could treat the subsequent optimization of such new features as a follow on project

Reply:
Updated to use Cow. <insert moo joke here>